packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java

```diff
@@ -597,8 +616,67 @@ static String interpretCircleOptions(Object o, CircleOptionsSink sink) {
     }
   }
 
+  static String interpretHeatmapOptions(Object o, HeatmapOptionsSink sink) {
+    final Map<?, ?> data = toMap(o);
+    final Object rawWeightedData = data.get("data");
+    if (rawWeightedData != null) {
+      sink.setWeightedData(toWeightedData(rawWeightedData));
+    }
+    final Object gradient = data.get("gradient");
+    if (gradient != null) {
+      sink.setGradient(toGradient(gradient));
+    }
+    final Object maxIntensity = data.get("maxIntensity");
+    if (maxIntensity != null) {
+      sink.setMaxIntensity(toDouble(maxIntensity));
+    }
+    final Object opacity = data.get("opacity");
+    if (opacity != null) {
+      sink.setOpacity(toDouble(opacity));
+    }
+    final Object radius = data.get("radius");
+    if (radius != null) {
+      sink.setRadius(toInt(radius));
+    }
+    final String heatmapId = (String) data.get("heatmapId");
+    if (heatmapId == null) {
+      throw new IllegalArgumentException("heatmapId was null");
+    } else {
+      return heatmapId;
+    }
+  }
+
+  static Map<String, Object> heatmapToJson(HeatmapTileProvider heatmap)
+      throws NoSuchFieldException, IllegalAccessException {
+    final Field dataField = HeatmapTileProvider.class.getDeclaredField("mData");
+    final Field gradientField = HeatmapTileProvider.class.getDeclaredField("mGradient");
+    final Field maxIntensityField =
+        HeatmapTileProvider.class.getDeclaredField("mCustomMaxIntensity");
+    final Field opacityField = HeatmapTileProvider.class.getDeclaredField("mOpacity");
+    final Field radiusField = HeatmapTileProvider.class.getDeclaredField("mRadius");
```

Comment:
I did this for the Android integration tests. The only alternative I see is to not have Android Integration tests for this code.

Reply:
If there are aspects of heatmaps that we can't validate with integration tests, then we should cover those with native unit tests instead (ideally some that are as close as possible to "end to end" within the native code path).

Reply:
If there are aspects of heatmaps that we can't validate with integration tests, then we should cover those with native unit tests instead (ideally some that are as close as possible to "end to end" within the native code path).

Reply:
There is no way to pull the heatmap data out of the maps without reflection regardless of the method in which we are testing

Reply:
Native unit tests can use dependency injection to replace the actual Google Maps SDK calls with a fake/mock object to validate the provided information, removing the need to read anything private out of the SDK.

The goal of plugin tests is to make sure that we are calling the SDK correctly in the plugin code, not that the SDK itself works correctly.

Reply:
Does using reflection for testing count as "using reflection in production"? Right now this test is a full integration test and any test without reflection would be inferior in the amount of code it covers.

Reply:
> Does using reflection for testing count as "using reflection in production"?

If the code is compiled by plugin clients and ends up in their application shipped to all of their users, then it's in production.

> Right now this test is a full integration test

I'm aware of that, but it's an integration test implemented using an approach that we would not land.

> and any test without reflection would be inferior in the amount of code it covers.

And vastly superior in terms of not being tightly coupled to the internal implementation details of code that we don't control and is therefore subject to change at any time that would break our tests.